### PR TITLE
Fix etcd cache

### PR DESCRIPTION
**Pull request by malletp**
### What does this PR do?

Fix ls function of etcd cache and improve performance when etcd cache is used with many minions.

### What issues does this PR fix or reference?

#55305
#55988

### Previous Behavior
Error message like
'''
SaltCacheError: There was an error reading the key, /salt_cache/minions/minion_id/mine/data: Not a directory : /salt_cache/minions/minion_id/mine
'''

Filtering minions with grains/pillar takes time when etcd cache is used with many minions.

### New Behavior

Error fixed and performance improved.

### Tests written?

No

### Commits signed with GPG?

Yes

### CODE REVIEW

File: salt/cache/etcd_cache.py [hunk 1]
```diff
@@ -119,6 +121,28 @@ def _init_client():
...
+    '''
+    global cache
+    if cache is None or cache['last_update'] < datetime.datetime.now() - datetime.timedelta(seconds=5):
+        _init_client()
+        etcd_key = '{0}/{1}'.format(path_prefix, bank.split('/')[0])
+        try:
+            data = client.read(etcd_key, recursive=True)
+            cache = {'last_update': datetime.datetime.now()}
+            cache['data'] = {}
+            for leave in data.leaves:
```
Comment (DmitryKuzmenko) [inline salt/cache/etcd_cache.py]: Could you please change `leave` to `leaf`?

**Review by DmitryKuzmenko — CHANGES_REQUESTED** [review]

**Review by DmitryKuzmenko — DISMISSED** [review]

**Comment by DmitryKuzmenko** [comment]
@malletp i've fixed the typo, don't care about that.
Are you able to write a test for your change please?

**Comment by DmitryKuzmenko** [comment]
@malletp also sorry I've missed the key point: please switch this PR to target to the `master` branch.

**Comment by DmitryKuzmenko** [comment]
@malletp Could you please update your PR onto the top of master branch and run the pre-commit script on your code to re-format it with `isort` and `black` we're using now.
```
pip install pre-commit
pre-commit install
FILES="$(git show --pretty="" --name-only master..HEAD)"
pre-commit run isort $FILES
pre-commit run black $FILES
```
next time you'll commit it will be ran automatically.

**Comment by malletp** [comment]
@DmitryKuzmenko, thanks' for the fix. 
Regarding the test, I don't find any existing test for the different supported cache module, and I have no idea how to add it.
And regarding the branch, I try to update the PR this week.

**Comment by DmitryKuzmenko** [comment]
@malletp thank you! Let's rebase this to the master as a first step then we'll check what to do with the tests. It should be pretty simple since we already have some cache unit tests.

**Comment by DmitryKuzmenko** [comment]
@malletp I'm sorry if it wasn't clear. It's not enough to just change the target branch. You have to rebase your commits to master locally using `git rebase -i --onto master 2018.3 fix-etcd-cache`. This command will rebase the only your commits to the top of master.
After that please install and run the pre-commit hook as I suggested. This will make your code clean.

**Review by Ch3LL — CHANGES_REQUESTED** [review]
Can you add test coverage for this change?

**Comment by Ch3LL** [comment]
@malletp did you see my review comments? Also there is a merge conflict and failing tests.

**Comment by Ch3LL** [comment]
any interest in pushing this PR to completion @malletp ?

**Comment by Ch3LL** [comment]
Closing due to inactivity. Please let me know if you want me to re-open or please submit a PR against the master branch with the requested changes.

**Pull request closed without merging**